draft of day 5 (#10)
UW-Madison-DataScience/geospatial-capstone: _episodes_rmd/06_day5.Rmd

---
title: "Day 6"
teaching: 0
exercises: 0
questions:
- "change"

> objectives:

Reviewer: I know this is still in draft form, but it might be a good to add in the objectives and questions posed in this capstone lesson.

- "change"
keypoints:
- "First key point. Brief Answer to questions. (FIXME)"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#source("../bin/chunk-options.R")
#knitr_fig_path("03-")
```

# Outline
- Addressing Minute questions (10-15 min)

> - Revisit your spatial question

Reviewer: What is the spatial question - is something from the previous capstone?

- Add point data from a non-spatial CSV file (thoughts:Use csv to add in points for each county (we would need to create these points ahead) - these points could be precipitation collection spots, weather forecasting locations, centroids?, major cities?)
- Crop raster data with vector data and output summary information (thoughts:Take raster data and crop with vector or point and vector and summarize stats.  Question for them to solve.  Mean temp diff for two cities?  Max precipitation of every county?)
- Thing to highlight: Using last question to also talk about pros/cons of raster vector data with answering questions like these
- Addressing final questions (10-15 min)

# Revisit your spatial question

On day 1 we brainstormed questions you could ask with geospatial data. You thought about questions you can answer with rater data, and separately, questions you could answer with vector data. Through this workshop you have practiced worked with spatial data and have the tools to answer your questions.

> In this final day we will implement tools you learned for combining rater and vector data and the types of questions you can answer when doing so.

Reviewer: small typo, raster, not rater ;)

# import point data from a csv file
WIcities <- read.csv("data/ne_10m_populated_places_Wisconsin.csv")

# check out the data
colnames(WIcities)

```

Recall from your lesson how to import point data and convert it to a spatial data layer in R.

When looking at the column names we see that there is a lot of data! To convert to a spatial points data layer, we need to know information about the coordinate system and the coordinates for each point.

The column names 'LATITUDE' and 'LONGITUDE' have what we need! 

Explain how we know about the coordinate system because of these coordinates.
Reference this website: https://spatialreference.org/ref/epsg/wgs-84/ 

> ## Solution: Load .csv point data?
>
> Text here describing what's in this box
>
>
> > ## Solution
> >
> > ```{r, echo=FALSE}
> > paste("This", "new", "template", "looks", "good")
> > ```
> {: .solution}

```{r convert-to-points}

# convert data framem to sf type of object by specifying coordinates (as xy) and the projection as a crs
WIcities_points <- st_as_sf(WIcities, coords=c('LONGITUDE','LATITUDE'), crs=crs(maxTemp_global))

# check class of object - should be "sf" data frame
class(WIcities_points)

```

> A critical part of working with raster and vector data together is that the dataasets are in the same coordinate system. 

Reviewer: small typo, dataasets

Reviewer 2: ```suggestion
A critical part of working with raster and vector data together is that the datasets are in the same coordinate system. 
```

Does it make sense? Are they the same?

```{r check-coordinate-systems}

crs(maxTemp_global)
crs(WIcities_points)

crs(maxTemp_global)
crs(WIcities_points)

extent(maxTemp_global)
extent(WIcities_points)

res(maxTemp_global)

> ```

Reviewer: I really like this part of the lesson looking at crs and extent - it re-emphasizes a lot of the stuff covered in the lessons!

Finally, let's subset our spatial data to only include the city points we are interested in. For the sample dataset, we are going to subset the data to only include the cities of Madison and Milwaukee.

If you are using your own data, you may or may not need to do this set.

```{r subset-vector-data}

WIcities_points$NAME

MM_points <- WIcities_points[WIcities_points$NAME=='Madison'|WIcities_points$NAME=='Milwaukee',]

```

# Crop raster data by vector locations

UPDATE. Langugae about cropping. Why do this. We will crop to WI cities, then plot to see the data better.

> 

Reviewer: Maybe introduce the `crop()` function first? What does it do, what are the arguments?

If you are working with your own data, ask yourself:
+ How do I want to summarize the data? For example, do I want a single mean value? A maximum? A distribution of all values?
+ What is the area I want to summarize? If it's a point, do you want a certain distance around the point? Or, if it's an area, you will include all of the raster values within that area.

One thing you'll need to consider is the converstion from degrees to meters. The conversions depends on where you are on the globe. For example, at the equator, 1 degree is 111 km. This distance decreases as you move closer to the poles. Here is an article: https://www.usgs.gov/faqs/how-much-distance-does-a-degree-minute-and-second-cover-your-maps?qt-news_science_products=0#qt-news_science_products 

```{r extract-raster-by-vector}

# I'm exepcting a distibution but am getting a single value
# does it have to do with the resolution and the buffer size?
df_maxTemp_MM <- as.data.frame(extract(x=maxTemp_global,df=TRUE,
                           y=MM_points,

>                            buffer=50000))

Reviewer: Can you elaborate more on what `buffer=50000` does and why we need it?

---

> title: "Day 6"

Reviewer 2: ```suggestion
title: "Day 5"
```

Wait, why is spatial resolution in units of time? 
This is because the spatial coordinates on a sphere are often expressed as degrees-minutes-seconds. 10 minutes is the biggest and 30 seconds is the smallest. An online conversion tool is helpful to understand these unit conversions(https://www.opendem.info/arc2meters.html)

For the capstone data, we want to use the largest spatial resolution of 10 minutes. This means each pixel represents about 13 km on the ground at the approximate latitude of Wisconsin. Within the "bio" variables, we will use the "bio5" variable that represents the maxiumum temperature of the warmest month.

> ## Solution
> ```{r load-raster-data}
> 
> # get worldclim data, variable "bio" with a 10 minute resolution
> # this line is downloading data
> # if you selected a resolution higher than 10 (with a res of 0.5,2.5, or 5) it may take a while to download
> worldClim_bio_global <- getData("worldclim", var="bio", res=10)
> 
> # use only the bio5 variable (BIO5 = Max Temperature of Warmest Month
> maxTemp_global <- worldClim_bio_global$bio5
> 
> # plot data
> plot(maxTemp_global)
> ```
{: .solution}

## Load vector data

> The capstone data includes the Natual Earth Populated Places dataset. We have provided the dataset for you as a ".csv" file in the Capstone data folder. By the end of this section you should have a simple feature ('sf') data frame type of object in your Global Environment.

Reviewer: ```suggestion
The capstone data includes the Natural Earth Populated Places dataset. We have provided the dataset for you as a ".csv" file in the Capstone data folder. By the end of this section you should have a simple feature ('sf') data frame type of object in your Global Environment.
```

> ## Solution
> ```{r crop-raster-by-vector}
> 
> maxTemp_WI <- crop(x = maxTemp_global, y = extent(WIcities_points),snap="out")
> maxTemp_WI_df <- as.data.frame(maxTemp_WI, xy=T)
> 
> ggplot() +
>   geom_raster(data = maxTemp_WI_df,aes(x = x, y = y, fill = bio5)) + 
>   geom_sf(data = MM_points,color="yellow",size=5) + 
>   coord_sf()
> ```
{: .solution}

# 4) Summarize raster data by vector locations
Now that we have our raster layer that contains the data we want to summarize, and the vector data that contains the locations indicating where we want to summarize the data, we can now put them together.

If you are working with your own data, ask yourself:

+ How do I want to summarize the data? For example, do I want a single mean value? A maximum? A distribution of all values?
+ What is the area I want to summarize? If it's a point, do you want a certain distance around the point? Or, if it's an area, you will include all of the raster values within that area.

## Extract pixel values

> For the capstone data, we want to extract all the pixel values within a distance buffer of the two city points. Since our data is "unprojected" and therefore the units are in latitude and longitude, the buffer distance will be in meters (hint: this information is in the Help file for the `extract` function!)

Reviewer: ```suggestion
For the capstone data, we want to extract all the pixel values within a distance buffer of the two city points. Since our data is "unprojected" and therefore the units are in latitude and longitude, the buffer distance will be in meters (hint: this information is in the Help file for the `extract()` function!)
```

> ## Geographic and projected unit reminder
>
> Remember, your raster is "unprojected" meaning that the pixels represent distances on a sphere of degrees, minutes, and seconds.

> > You'll need to consider is the converstion from degrees to meters. The conversions depends on where you are on the globe. For example, at the equator, 1 degree is 111 km. This distance decreases as you move closer to the poles.

Author: ```suggestion
> You'll need to consider is the conversation from degrees to meters. The conversions depends on where you are on the globe. For example, at the equator, 1 degree is 111 km. This distance decreases as you move closer to the poles. Try setting the buffer to 50 km or 50,000 meters. This is something you can play around with as you explore the data.
```

>
{: .callout}

> ## Solution
> ```{r extract-raster-by-vector}
> 
> # extract pixel values from the raster based on the Milwaukee and Madison points
> # using a buffer size of 50,000 (50 km around the each point) because the units are in meters
> maxTemp_MM <- extract(x=maxTemp_global,df=TRUE,
>                            y=MM_points,

> >                            buffer=50000)

Reviewer: Should the prompt state to set the `buffer=50000` or are you allowing the learners to explore different buffers?

# 2) Load libraries and import data
For raster data, we use functions in the `raster` package. For vector data, you will use functions in the `sf` package. In addition, because we will visualize our data, we will use the plotting package `ggplot2`.

> ## Solution
> ```{r load package, message=FALSE}
> 
> library(raster)
> library(sf)
> library(ggplot2)
> ```
{: .solution}

## Load raster data
The capstone data includes the World Climate Data or "WorldClim". We will use your coding skills to access the worldclim dataset from the `raster` package, crop it to our extent of interest, and summarize the data. By the end of this section you should have your raster data as a 'Formal class RasterLayer' in your Global Environment

> ## World Climate Data
>
> You can learn about and download the WorldClim from the website: https://www.worldclim.org/data/worldclim21.html 
> For an example of how to use the `raster` package to access the WorldClim (raster) data and GADM (vector) data, visit this tutorial: https://www.gis-blog.com/r-raster-data-acquisition/ 
> 
>

> > {: .source}

Reviewer 2: ```suggestion
{: .callout}
```

> > {: .source}

Author: ```suggestion

```

---
title: "Day 5"
questions:
- "How can I summarize raster data based on vector locations"
objectives:
- "Import .csv files containing x,y coordinate locations and convert to a spatial object"
- "Crop a raster to the extent of a vector layer"
- "Extract values from a raster that correspond to a vector file overlay"
keypoints:
- "Convert a data frame to an `sf` object using the `st_as_sf()` function."
- "Use the `crop()` and `extract()` functions to get raster data of spatial subsets."
- "Remember to check if your data align spatially by looking at the projection, extent, and units"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#source("../bin/chunk-options.R")
#knitr_fig_path("03-")
```

# Outline

- Addressing Minute questions (10-15 minutes)

> - Capstone lesson

Reviewer 3: Does this item need a time estimate?

Author: oh yeah, that would make sense 👍 
```suggestion
- Capstone lesson (30-40 minutes)
```

    1. Revisit your spatial question
    2. Load libraries and import data
    3. Crop raster with a vector
    4. Summarize raster data by vector locations
- Keypoints and final questions (10-15 minutes)

# 1) Revisit your spatial question

On day 1 we brainstormed questions you could ask with geospatial data. You thought about questions you can answer with raster data, and separately, questions you could answer with vector data. Through this workshop you have practiced working with spatial data and have the tools to answer your question.

In this final day we will implement tools you learned for combining raster and vector data and the types of questions you can answer when doing so.

A spatial question may look like the following:

> + where does something occur?
> + how does something change across space or time?
> + what are the differences between locations?

Reviewer 3: ```suggestion
+ Where does something occur?
+ How does something change across space or time?
+ What are the differences between locations?
```

# 2) Load libraries and import data
For raster data, we use functions in the `raster` package. For vector data, you will use functions in the `sf` package. In addition, because we will visualize our data, we will use the plotting package `ggplot2`.

> ## Solution
> ```{r load package, message=FALSE}
> 
> library(raster)
> library(sf)
> library(ggplot2)
> ```
{: .solution}

## Load raster data
The capstone data includes the World Climate Data or "WorldClim". We will use your coding skills to access the worldclim dataset from the `raster` package, crop it to our extent of interest, and summarize the data. By the end of this section you should have your raster data as a 'Formal class RasterLayer' in your Global Environment

> ## World Climate Data
>
> You can learn about and download the WorldClim from the website: https://www.worldclim.org/data/worldclim21.html 
> For an example of how to use the `raster` package to access the WorldClim (raster) data and GADM (vector) data, visit this tutorial: https://www.gis-blog.com/r-raster-data-acquisition/ 
> 

> >
> > {: .source}

Reviewer 3: ```suggestion
```

I don't think you need the `{: .source}` tag here if you don't have source code in the callout.

Reviewer 3: @sstevens2 can you confirm?

Author: oh yeah you could be right. I'll leave this comment here for now.

> ## Solution
> ```{r load-raster-data}
> 
> # get worldclim data, variable "bio" with a 10 minute resolution
> # this line is downloading data
> # if you selected a resolution higher than 10 (with a res of 0.5,2.5, or 5) it may take a while to download
> worldClim_bio_global <- getData("worldclim", var="bio", res=10)
> 
> # use only the bio5 variable (BIO5 = Max Temperature of Warmest Month
> maxTemp_global <- worldClim_bio_global$bio5
> 
> # plot data
> plot(maxTemp_global)
> ```
{: .solution}

## Load vector data
The capstone data includes the Natual Earth Populated Places dataset. We have provided the dataset for you as a ".csv" file in the Capstone data folder. By the end of this section you should have a simple feature ('sf') data frame type of object in your Global Environment.

> ## Natural Earth Data
>
> You can learn about and download the Natural Earth dataset from the website: https://www.naturalearthdata.com/
> You can also download Natural Earth data directly into R using the [`rnaturalearth` package] (https://cran.r-project.org/web/packages/rnaturalearth/README.html)
> 

> > {: .source}

Reviewer 3: ```suggestion
```

Same as above.

{: .callout}

> ## Solution
> ```{r load-city-point-data}
> 
> # import point data from a csv file
> WIcities <- read.csv("../data/ne_10m_populated_places_Wisconsin.csv")
> 
> # check out the data
> colnames(WIcities)
>```
{: .solution}

Recall from your lesson how to import point data and convert it to a spatial data layer in R.

When looking at the column names we see that there is a lot of data! To convert to a spatial points data layer, we need to know information about the coordinate system and the coordinates for each point.

> The column names 'LATITUDE' and 'LONGITUDE' have what we need! 

Reviewer 3: ```suggestion
The column names `LATITUDE` and `LONGITUDE` have what we need! 
```

> ## Solution
> ```{r convert-to-points}
> 
> # convert data frame to sf type of object by specifying coordinates (as xy) and the projection as a crs
> WIcities_points <- st_as_sf(WIcities, coords=c('LONGITUDE','LATITUDE'), crs=crs(maxTemp_global))
> 
> # check class of object - should be "sf" data frame
> class(WIcities_points)
> ```
{: .solution}

A critical part of working with raster and vector data together is that the datasets are in the same coordinate system. 

Does it make sense? Are they the same? You can use the functions in the `raster` package to check out your data.

```{r check-coordinate-systems}

crs(maxTemp_global)
crs(WIcities_points)

extent(maxTemp_global)
extent(WIcities_points)

res(maxTemp_global)

```

Finally, let's subset our spatial data to only include the city points we are interested in. For the sample dataset, we are going to subset the data to only include the cities of Madison and Milwaukee.

> If you are using your own data, you may or may not need to do this subset.

Reviewer 3: ```suggestion
If you are using your own data, you may or may not need to do this subset or you might want to choose different locations.
```

> ## Solution
> ```{r crop-raster-by-vector}
> 
> maxTemp_WI <- crop(x = maxTemp_global, y = extent(WIcities_points),snap="out")
> maxTemp_WI_df <- as.data.frame(maxTemp_WI, xy=T)
> 
> ggplot() +
>   geom_raster(data = maxTemp_WI_df,aes(x = x, y = y, fill = bio5)) + 
>   geom_sf(data = MM_points,color="yellow",size=5) + 
>   coord_sf()
> ```
{: .solution}

# 4) Summarize raster data by vector locations
Now that we have our raster layer that contains the data we want to summarize, and the vector data that contains the locations indicating where we want to summarize the data, we can now put them together.

If you are working with your own data, ask yourself:

+ How do I want to summarize the data? For example, do I want a single mean value? A maximum? A distribution of all values?
+ What is the area I want to summarize? If it's a point, do you want a certain distance around the point? Or, if it's an area, you will include all of the raster values within that area.

## Extract pixel values

> For the capstone data, we want to extract all the pixel values within a distance buffer of the two city points. Since our data is "unprojected" and therefor the units are in lat/long, the buffer distance will be in meters (hint: this information is in the Help file for the `extract` function!)

Reviewer 3: ```suggestion
For the capstone data, we want to extract all the pixel values within a distance buffer of the two city points. Since our data is "unprojected" and therefore the units are in latitude and longitude, the buffer distance will be in meters (hint: this information is in the Help file for the `extract` function!)
```

> ## Geographic and projected unit reminder
>
> Remember, your raster is "unprojected" meaning that the pixels represent distances on a sphere of degrees, minutes, and seconds.
> You'll need to consider is the converstion from degrees to meters. The conversions depends on where you are on the globe. For example, at the equator, 1 degree is 111 km. This distance decreases as you move closer to the poles.
>

> >
> > {: .source}

Reviewer 3: ```suggestion
```

Same as above
